### ISSUE TITLE
False positives with empty check

### ISSUE THREAD

**Issue by rliebz** [Antonboom/testifylint#6]
A check for `Greater(len(elems), 1)` is not equivalent to `NotEmpty`. In cases where the list is exactly length 1, the original test would fail, while the suggestion would succeed.

The same logical error is present for `Less(1, len(elems))` and `NotEmpty`. 

Simple reproduction:

```go
func TestExample(t *testing.T) {
	elems := []int{1}
	assert.Greater(t, len(elems), 1)
	assert.Less(t, 1, len(elems))
}
```

Expected: No error
Actual: `empty: use assert.NotEmpty` for each assert

If we actually apply the suggested fix:

```go
func TestExampleFix(t *testing.T) {
	elems := []int{1}
	assert.NotEmpty(t, elems)
}
```

The "fixed" test passes, while the original test failed.

**Comment by Antonboom** [comment]
@rliebz, hi!

Thank u for report.

While we are awaiting of golangci-lint patch you could use as workaround:

```go
func TestExample(t *testing.T) {
	elems := []int{1}
	assert.Greater(t, len(elems), 1) //nolint:testifylint // https://github.com/Antonboom/testifylint/issues/6
}
```

Or change set of enabled checkers via [config](https://golangci-lint.run/usage/linters/#testifylint).

P.S. What is your feedback about other checks? No issues more?

**Comment by rliebz** [comment]
Appreciate the quick fix here! This tool caught a lot of very valid issues in our project, so I've been pretty happy with it overall. 

The main other issue I've had is that `require-error` throws up a lot of false positives based on some of our error checks.

One really prominent example is based on this snippet from `testing.T.FailNow`:

```go
// FailNow must be called from the goroutine running the
// test or benchmark function, not from other goroutines
// created during the test. Calling FailNow does not stop
// those other goroutines.
```

There are lots of situations in which even if `require` would have been preferable, it is not allowed, because it's not being called from the main goroutine. 

Another example is that certain tests have much more nuanced error checking. Here's a simplified example:

```go
func TestNewStatusError(t *testing.T) {
	err := NewStatusError("oh no", StatusNotFound)
	assert.EqualError(t, err, "status error: oh no")

	var statusErr StatusError
	if assert.ErrorAs(t, err, &statusErr) {
		assert.Equal(t, StatusNotFound, statusErr.StatusID())
	}

	// more test logic...
}
```

The first check is simply for an error message, and the second correctly guards the code by using an `if` statement. The test correctly handles cases where the error is nil, but gets two warnings for the `assert.EqualError` and `assert.ErrorAs` calls. Maybe it makes sense to add `require.Error(t, err)` at the top, but even if I do, I still get the false positives on subsequent lines. 

And quite simply, there are cases where an error might be relevant to check for but has nothing to do with whether test execution should continue. It's hard to demonstrate this with a simplified code snippet, but there are a few places in the code where I've found this pop up. As a descriptive example, issuing a ping with a database pool to prove it was set up correctly, but for a test whose purpose is to prove that the database was set up correctly, and therefore is not guarding the rest of the function from executing in case of failure. 

After those handful of false positives, I decided it wasn't worth keeping enabled. I ended up not filing a separate issue because it seems like the way I write tests disagrees with the premise of the rule, even if most of the time it gives good advice. I'm not sure there's a way to lint that without false positives. 

The other challenge I have is that it's a bit annoying to disable a single rule. My config looks like this:
```yaml
  testifylint:
    enable:
      # Disable require-error; it disallows valid use
      - bool-compare
      - compares
      - empty
      - error-is-as
      - error-nil
      - expected-actual
      - float-compare
      - len
      - suite-dont-use-pkg
      - suite-extra-assert-call
      - suite-thelper
```

When I think ideally it would look more like this:
```yaml
  testifylint:
    enable-all: true
    disable:
      - require-error
```

**Comment by Antonboom** [comment]
Thank you for detailed comment. At the moment I am working on `require-error` false positives minimisation and I agree with highlighted issues.

> it is not allowed, because it's not being called from the main goroutine.

For me any assertion in goroutine is anti-pattern because `*testing.T` type is not thread-safe.
go vet's [testinggoroutine](https://pkg.go.dev/golang.org/x/tools/go/analysis/passes/testinggoroutine) in this topic too.

> When I think ideally it would look more like this:

Initially it was exactly so, but I was worried about the thoughtless enabling of linters (as happens in `golangci-lint`). I think I'll return a more detailed configuration back 👌

**Comment by Antonboom** [comment]
Some offtop:

Analyze code like this:
```go
if assert.ErrorAs(t, err, &statusErr) {
    assert.Equal(t, StatusNotFound, statusErr.StatusID())
}
```
or this 
```go
err := oex.MountVolume(waitForAttachTimeout, volumeToMount, asw, false)
// Assert
if assert.Error(t, err) {
    assert.Contains(t, err.Error(), tc.expectedErrMsg)
}
```

I came to the conclusion that this is an incorrect code structure because it hides a bugs.
The more right way:
```go
require.ErrorAs(t, err, &statusErr) {
assert.Equal(t, StatusNotFound, statusErr.StatusID())
```
```go
if tc.expectedErrMsg != nil {
    require.Error(t, err)
    assert.ErrorContains(t, err, tc.expectedErrMsg)
}
```
etc.

**Comment by rliebz** [comment]
From https://pkg.go.dev/testing#T:

>A test ends when its Test function returns or calls any of the methods FailNow, Fatal, Fatalf, SkipNow, Skip, or Skipf. Those methods, as well as the Parallel method, must be called only from the goroutine running the Test function.
>
>The other reporting methods, such as the variations of Log and Error, may be called simultaneously from multiple goroutines. 

`*testing.T` as a type is not thread unsafe—the issue is that a test cannot be terminated from inside of another test goroutine. Which effectively means `require` can never be called from another test goroutine, while `assert` can. Of course all the challenges of concurrent programming still apply (e.g., making sure a goroutine terminates before the end of the main test goroutine), but `t.Error`/`t.Log`/`assert.*` are all thread safe.

**Comment by Antonboom** [comment]
>  is that a test cannot be terminated from inside of another test goroutine

Yes, and the reason is using of `runtime.Goexit` inside `FailNow`.
you don't have to explain this to me :)

I just remember some guard wrappers around `*testing.T`, but no I see that the methods contain locks since the creation of this package :) Confused, aha

About `no require in goroutines` – looks like new checker 🤔 👍
